### PR TITLE
Feature/merge logging

### PR DESCRIPTION
**Pull request by TurtleARM**

### CODE REVIEW

File: Logging.Tests/LoggingTest.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            
+            var result = appender.GetEvents();
+            Assert.IsTrue(result.Any());
+            Assert.IsTrue(result[0].MessageObject.Equals(expectedLog));
+        }
+        
+        [Test]
+        public void TestCompletedRequestLog()
+        {
+            var expectedLog =
```
Comment (dibericky) [inline Logging.Tests/LoggingTest.cs:43]: Should logs follow are guidelines?
https://docs.mia-platform.eu/development_suite/monitoring-dashboard/dev_ops_guide/log/

Fieldnames looks different from what specified in that guidelines

Comment (TurtleARM) [reply Logging.Tests/LoggingTest.cs:43]: I followed this example to compose log fields (Mandatory Logs): https://github.com/mia-platform/lc39/blob/master/lib/custom-logger.js#L43

Looking at the guidelines, fields have to be in lowercase, and the mandatory field "request_id" has to be added.

File: Logging.Tests/LoggingTest.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                            bytes = Bytes
+                        },
+                        statusCode = StatusCode
+                    }
+                },
+                url = new Url 
+                {
+                    path = Path,
+                },
+                userAgent = new UserAgent
```
Comment (fredmaggiowski) [inline Logging.Tests/LoggingTest.cs:73]: `userAgent` key has been recently moved below `request`.

File: Logging.Tests/LoggingTest.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                host = new Host
+                {
+                    hostname = Hostname,
+                    ip = Ip
+                },
+                responseTime = ResponseTime
+            };
+        }
+    }
+}
```
Comment (fredmaggiowski) [inline Logging.Tests/LoggingTest.cs:86]:
```suggestion
}

```

File: Logging/Entities/Body.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Logging.Entities
+{
+    public class Body
+    {
+        public long bytes { get; set; }
+    }
+}
```
Comment (fredmaggiowski) [inline Logging/Entities/Body.cs:7]: Please always add new lines.

I don't understand why it is necessary to define all these classes for logging properties: can't we use a map?

```suggestion
}

```

File: Logging/Entities/CPRequest.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Logging.Entities
+{
+    public class CpRequest
+    {
+        public string method { get; set; }
+    }
+}
```
Comment (fredmaggiowski) [inline Logging/Entities/CPRequest.cs:7]:
```suggestion
}

```

File: Logging/Entities/CPResponse.cs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Logging.Entities
+{
+    public class CpResponse
+    {
+        public int statusCode { get; set; }
+        public Body body { get; set; }
+    }
+}
```
Comment (fredmaggiowski) [inline Logging/Entities/CPResponse.cs:8]:
```suggestion
}

```

File: Logging/Entities/Host.cs [hunk 7]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Logging.Entities
+{
+    public class Host
+    {
+        public string hostname  { get; set; }
+        public string ip  { get; set; }
+    }
+}
```
Comment (fredmaggiowski) [inline Logging/Entities/Host.cs:8]:
```suggestion
}

```

File: Logging/LoggingUtility.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+using Microsoft.AspNetCore.Http;
+using Newtonsoft.Json;
+
+namespace Logging
+{
+    public class LoggingUtility
+    {
+        private static readonly ILog Logger = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
+
+        private static void LogWithLevel(LogLevels level, string message)
```
Comment (fredmaggiowski) [inline Logging/LoggingUtility.cs:14]: So the developer has to call `LoggingUtility.LogWithLevel`? 

Does not seem like a clean interface, Why not: `Logger.Info(..)`, `Logger.Trace(..)`, etc

File: docs/README.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
+# Logging
```
Comment (fredmaggiowski) [inline docs/README.md:1]: Is this page linked in the README?

**Review by fredmaggiowski — CHANGES_REQUESTED** [review]

File: docs/README.md (hunk 9 above)
Comment (fabio-grieco) [reply docs/README.md]: Linked

File: Logging.Tests/LoggingTest.cs (hunk 2 above)
Comment (fabio-grieco) [reply Logging.Tests/LoggingTest.cs]: Ok, moved

File: Logging.Tests/LoggingTest.cs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            var expectedLog =
+                $@"{{""level"":{Level},""time"":{Time},""reqId"":{ReqId},""http"":{{""request"":{{""method"":""{HttpRequestMethod}"",""userAgent"":{{""original"":""{Original}""}}}},""response"":{{""statusCode"":{StatusCode},""body"":{{""bytes"":{Bytes}}}}}}},""url"":{{""path"":""{Path}""}},""host"":{{""hostname"":""{Hostname}"",""ip"":""{Ip}""}},""responseTime"":{ResponseTime}.0}}";
+            var mockRequest = BuildCompletedRequestLog();
+
+            var appender = new log4net.Appender.MemoryAppender();
+            BasicConfigurator.Configure(appender);
+            LoggingUtility.LogRequest(mockRequest);
+
+            var result = appender.GetEvents();
+            Assert.IsTrue(result.Any());
```
Comment (fabio-grieco) [inline Logging.Tests/LoggingTest.cs:35]: @TurtleARM all the tests under `Logging.Tests' were being ignored due to a missing test adapter. I included the adapter and ran all tests, now this test is failing on this assertion but I can't figure out how the appender should receive events.
image

**Review by fabio-grieco — CHANGES_REQUESTED** [review]

File: Logging/LoggingUtility.cs (hunk 8 above)
Comment (TurtleARM) [reply Logging/LoggingUtility.cs:14]: The idea was to replicate `req.log.{level}` from fastify, where 
 `req.log.info({ statusCode: 200 }, 'Notify sent!!!!')`
would produce the following output:
`{"level":30,"time":1601367510,"pid":7956,"hostname":"x","reqId":1,"statusCode":200,"msg":"Notify sent!!!!","v":1}` (from Node.js-PrePost-Microservice-Example)

The developer could also instantiate an ILog logger and then use the classic interface:

```
public class HelloWorldController : CpController
{
    private readonly ILog _logger = LogManager.GetLogger(typeof(HelloWorldController));
    [HttpPost]
    [Route("hello")]
    public string LoggingTest([FromBody] Hello hello)
    {
        // Fastify-like log message
        LoggingUtility.LogMessage(HttpContext.Request, LogLevels.Trace, hello, "testMessage");
        // Classic log4net log
        _logger.Trace("testTrace");
    }
}
```

File: Logging.Tests/LoggingTest.cs (hunk 10 above)
Comment (TurtleARM) [reply Logging.Tests/LoggingTest.cs]: Looks like MemoryAppender has some problems with github Actions, I changed the test logic to verify that the logging functions get called with the right parameters

File: Logging/RequestResponseLoggingMiddleware.cs [hunk 11]
```diff
@@ -12,41 +13,48 @@ public class RequestResponseLoggingMiddleware
...
-        
+
         public RequestResponseLoggingMiddleware(RequestDelegate next)
         {
             _next = next;
         }
         
         public async Task Invoke(HttpContext context)
-        {
+        {
```
Comment (fredmaggiowski) [inline Logging/RequestResponseLoggingMiddleware.cs:23]:
```suggestion
        {
```

File: docs/Logging.md [hunk 12]
```diff
@@ -14,19 +23,16 @@ By default the library will generate a log for each request, representing both t
...
     else
     {
-        // Log at 'Warn' level with custom object 'hello'
-        LoggingUtility.LogMessage(HttpContext.Request, LogLevels.Warn, null, "custom message 3");
+        Logger.Warn(HttpContext.Request, "bar");
     }
-    return "Goodbye, world";
-}
+    return "Goodbye, world!";
+}
```
Comment (fredmaggiowski) [inline docs/Logging.md:36]:
```suggestion
}
```

**Review by fredmaggiowski — APPROVED** [review]

**Review by fredmaggiowski — APPROVED** [review]

**Pull request merged**